src/EFCore.Analyzers/InternalUsageDiagnosticAnalyzer.cs

```
@@ -40,84 +42,244 @@ public class InternalUsageDiagnosticAnalyzer : DiagnosticAnalyzer
         }
 
+        private static void ReportDiagnostic(OperationAnalysisContext context, object messageArg)
+            => context.ReportDiagnostic(Diagnostic.Create(_descriptor, NarrowDownSyntax(context.Operation.Syntax).GetLocation(), messageArg));
+
+        private static void ReportDiagnostic(SymbolAnalysisContext context, SyntaxNode syntax, object messageArg)
+            => context.ReportDiagnostic(Diagnostic.Create(_descriptor, NarrowDownSyntax(syntax).GetLocation(), messageArg));
+
+        /// <summary>
+        ///     Given a syntax node, pattern matches some known types and returns a narrowed-down node for the type syntax which
+        ///     should be reported in diagnostics.
+        /// </summary>
+        private static SyntaxNode NarrowDownSyntax(SyntaxNode syntax)
+            => syntax switch
+            {
+                CSharpSyntax.InvocationExpressionSyntax s
+                    when s.Expression is CSharpSyntax.MemberAccessExpressionSyntax memberAccessSyntax
+                    => memberAccessSyntax.Name,
+                CSharpSyntax.MemberAccessExpressionSyntax s => s.Name,
+                CSharpSyntax.ObjectCreationExpressionSyntax s => s.Type,
+                CSharpSyntax.PropertyDeclarationSyntax s => s.Type,
+                CSharpSyntax.VariableDeclaratorSyntax declarator
+                    => declarator.Parent is CSharpSyntax.VariableDeclarationSyntax declaration
+                        ? declaration.Type
+                        : (SyntaxNode)declarator,
+                CSharpSyntax.TypeOfExpressionSyntax s => s.Type,
+
+                VBSyntax.InvocationExpressionSyntax s
+                    when s.Expression is VBSyntax.MemberAccessExpressionSyntax memberAccessSyntax
+                    => memberAccessSyntax.Name,
+                VBSyntax.MemberAccessExpressionSyntax s => s.Name,
+                VBSyntax.ObjectCreationExpressionSyntax s => s.Type,
+                VBSyntax.TypeOfExpressionSyntax s => s.Type,
```

It seems it would be better if you have different language specific analyzer assemblies for language specific code. Otherwise, your analyzer will cause a C# only solution to load all VB Roslyn assemblies and vice versa, which has an unnecessary assembly load overhead.

---

That's a good point - even if the assembly load overhead would only apply at build time...

The problem is that our analyzers package - Microsoft.EntityFrameworkCore.Analyzers - isn't typically referenced directly by user, but is brought in transitively as a dependency of Microsoft.EntityFrameworkCore. I know that ASP.NET also have a similar pattern, do you have any guidance on this, or any particular reason why the additional VB assembly load should be avoided?

If this is a big issue, we can drop the VB dependency altogether - it's only used to produce better/narrower diagnostic locations - the warnings themselves would still appear as we're using language-independent APIs here.

---

Normally the way this would be done is that your analyzer package will have 3 assemblies - Core, C# specific and VB specific. Then the MSBuild/SDK machinery will ensure that C# projects have analyzer references to first two and VB projects to the first and third. This way if someone has a C# only solution or a VB only solution, they do not get penalized by loading non-required language Roslyn assemblies (which can be quite a few as most of Roslyn layers are divided the same way). @jmarolf can probably help with setting up/guiding on the SDK pieces.

---

Oh nice, was not aware of this... Any pointer to an analyzer package which does this would be sufficient!

---

@jmarolf did this work for the `Microsoft.CodeAnalysis.NetAnalyzers` nuget package that we now insert into the .NET SDK. He can give pointers on how you can match his apporoach.

---

Scoping this out to #22085

---

> Otherwise, your analyzer will cause a C# only solution to load all VB Roslyn assemblies and vice versa, which has an unnecessary assembly load overhead.

Note that it's slightly worse than this implies. Csc does not provide the VB assemblies, and vbc does not provide the C# assemblies. Command line use of an analyzer that references both will work if and only if the shared compiler server VBCSCompiler.exe is used, but that will occasionally fail to process a request and silently fall back to the language-specific executables, at which point the analyzers will _silently_ fail to run.